File: gateway_service/routes/account.py
"""Маршруты Gateway → Account Service (банковские счета)."""

from uuid import UUID

from fastapi import APIRouter, Depends, Request, status

from shared import schemas
from gateway_service.helpers import forward_request
from gateway_service.middleware import session_token_scheme

router = APIRouter(
	prefix="/accounts",
	tags=["accounts"],
	dependencies=[Depends(session_token_scheme)],
)


@router.post(
	"",
	response_model=schemas.AccountMessageResponse,
	status_code=status.HTTP_201_CREATED,
	summary="Открыть новый счёт",
)
async def open_account(
	payload: schemas.OpenAccountRequest,
	request: Request,
):
	"""Создаёт банковский счёт указанного типа и валюты для текущего пользователя."""
	data = await forward_request(
		request,
		"POST",
		"/accounts",
		payload.model_dump(mode="json"),
		service="account",
	)
	return data


@router.get(
	"",
	response_model=schemas.AccountListResponse,
	status_code=status.HTTP_200_OK,
	summary="Список счетов",
)
async def list_accounts(request: Request):
	"""Возвращает все счета текущего пользователя."""
	data = await forward_request(
		request,
		"GET",
		"/accounts",
		service="account",
	)
	return data


@router.get(
	"/{account_id}",
	response_model=schemas.AccountResponse,
	status_code=status.HTTP_200_OK,
	summary="Детали счёта",
)
async def get_account(account_id: UUID, request: Request):
	"""Возвращает данные конкретного счёта."""
	data = await forward_request(
		request,
		"GET",
		f"/accounts/{account_id}",
		service="account",
	)
	return data


@router.post(
	"/{account_id}/close",
	response_model=schemas.AccountMessageResponse,
	status_code=status.HTTP_200_OK,
	summary="Закрыть счёт",
)
async def close_account(account_id: UUID, request: Request):
	"""Закрывает банковский счёт. Баланс должен быть 0."""
	data = await forward_request(
		request,
		"POST",
		f"/accounts/{account_id}/close",
		service="account",
	)
	return data


@router.post(
	"/{account_id}/freeze",
	response_model=schemas.AccountMessageResponse,
	status_code=status.HTTP_200_OK,
	summary="Заморозить счёт",
)
async def freeze_account(account_id: UUID, request: Request):
	"""Замораживает счёт. Исходящие операции блокируются, входящие — разрешены."""
	data = await forward_request(
		request,
		"POST",
		f"/accounts/{account_id}/freeze",
		service="account",
	)
	return data


@router.post(
	"/{account_id}/unfreeze",
	response_model=schemas.AccountMessageResponse,
	status_code=status.HTTP_200_OK,
	summary="Разморозить счёт",
)
async def unfreeze_account(account_id: UUID, request: Request):
	"""Размораживает счёт. Доступно только если заморозка инициирована пользователем."""
	data = await forward_request(
		request,
		"POST",
		f"/accounts/{account_id}/unfreeze",
		service="account",
	)
	return data

File: gateway_service/routes/account.go
package routes

import (
	"fmt"
	"net/http"

	"github.com/labstack/echo/v4"

	"gateway_service/proxy"
)

// AccountHandler обрабатывает маршруты account_service.
type AccountHandler struct {
	Proxy  *proxy.ServiceClients
	APIKey string
}

// RegisterAccountRoutes регистрирует маршруты банковских счетов.
func (h *AccountHandler) RegisterAccountRoutes(e *echo.Echo) {
	e.POST("/accounts", h.openAccount)
	e.GET("/accounts", h.listAccounts)
	e.GET("/accounts/:account_id", h.getAccount)
	e.POST("/accounts/:account_id/close", h.closeAccount)
	e.POST("/accounts/:account_id/freeze", h.freezeAccount)
	e.POST("/accounts/:account_id/unfreeze", h.unfreezeAccount)
}

// openAccount godoc
// @Summary     Открыть новый счёт
// @Description Создаёт банковский счёт указанного типа и валюты для текущего пользователя.
// @Tags        accounts
// @Security    SessionToken
// @Accept      json
// @Produce     json
// @Param       payload body schemas.OpenAccountRequest true "Тип и валюта счёта"
// @Success     201 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Failure     409 {object} map[string]string
// @Router      /accounts [post]
func (h *AccountHandler) openAccount(c echo.Context) error {
	body, _ := readBody(c)
	return h.Proxy.ForwardRaw(c, http.MethodPost, "/accounts", body, "account", h.APIKey)
}

// listAccounts godoc
// @Summary     Список счетов
// @Description Возвращает все счета текущего пользователя.
// @Tags        accounts
// @Security    SessionToken
// @Produce     json
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Router      /accounts [get]
func (h *AccountHandler) listAccounts(c echo.Context) error {
	return h.Proxy.ForwardRaw(c, http.MethodGet, "/accounts", nil, "account", h.APIKey)
}

// getAccount godoc
// @Summary     Детали счёта
// @Description Возвращает данные конкретного счёта.
// @Tags        accounts
// @Security    SessionToken
// @Produce     json
// @Param       account_id path string true "UUID счёта"
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Failure     404 {object} map[string]string
// @Router      /accounts/{account_id} [get]
func (h *AccountHandler) getAccount(c echo.Context) error {
	accountID := c.Param("account_id")
	path := fmt.Sprintf("/accounts/%s", accountID)
	return h.Proxy.ForwardRaw(c, http.MethodGet, path, nil, "account", h.APIKey)
}

// closeAccount godoc
// @Summary     Закрыть счёт
// @Description Закрывает банковский счёт. Баланс должен быть 0.
// @Tags        accounts
// @Security    SessionToken
// @Produce     json
// @Param       account_id path string true "UUID счёта"
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Failure     409 {object} map[string]string
// @Router      /accounts/{account_id}/close [post]
func (h *AccountHandler) closeAccount(c echo.Context) error {
	accountID := c.Param("account_id")
	path := fmt.Sprintf("/accounts/%s/close", accountID)
	return h.Proxy.ForwardRaw(c, http.MethodPost, path, nil, "account", h.APIKey)
}

// freezeAccount godoc
// @Summary     Заморозить счёт
// @Description Замораживает счёт. Исходящие операции блокируются, входящие — разрешены.
// @Tags        accounts
// @Security    SessionToken
// @Produce     json
// @Param       account_id path string true "UUID счёта"
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Router      /accounts/{account_id}/freeze [post]
func (h *AccountHandler) freezeAccount(c echo.Context) error {
	accountID := c.Param("account_id")
	path := fmt.Sprintf("/accounts/%s/freeze", accountID)
	return h.Proxy.ForwardRaw(c, http.MethodPost, path, nil, "account", h.APIKey)
}

// unfreezeAccount godoc
// @Summary     Разморозить счёт
// @Description Размораживает счёт. Доступно только если заморозка инициирована пользователем.
// @Tags        accounts
// @Security    SessionToken
// @Produce     json
// @Param       account_id path string true "UUID счёта"
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Router      /accounts/{account_id}/unfreeze [post]
func (h *AccountHandler) unfreezeAccount(c echo.Context) error {
	accountID := c.Param("account_id")
	path := fmt.Sprintf("/accounts/%s/unfreeze", accountID)
	return h.Proxy.ForwardRaw(c, http.MethodPost, path, nil, "account", h.APIKey)
}
